pkg/controller/external.go

```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(xpresource.IgnoreNotFound(err), "cannot check if resource exists")
+	}
+	if res.IsDestroying || res.IsApplying {
+		return managed.ExternalObservation{
+			ResourceExists:   true,
+			ResourceUpToDate: true,
+		}, nil
+	}
+	// After a successful observation, we now have a state to consume.
+	// We will consume the state by:
+	// - returning "sensitive attributes" as connection details
+	// - setting external name annotation, if not set already, from <id> attribute
+	// - late initializing "spec.forProvider" with "attributes"
+	// - setting observation at "status.atProvider" with "attributes"
+	// - storing base64encoded "tfstate" as an annotation
+
+	// No tfcli operation was in progress, our blocking observation completed
+	// successfully, and we have an observation to consume.
+	attr := map[string]interface{}{}
+	if err := json.JSParser.Unmarshal(res.State.GetAttributes(), &attr); err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, "cannot unmarshal state attributes")
+	}
+	if err := tr.SetObservation(attr); err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, "cannot set observation")
+	}
+
+	// TODO(hasan): Handle late initialization
+	lateInited, err := lateInitializeAnnotations(tr, attr, string(res.State.GetPrivateRaw()))
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, "cannot late initialize annotations")
+	}
+
+	// During creation (i.e. apply), Terraform already waits until resource is
+	// ready. So, I believe it would be safe to assume it is available if create
+	// step completed (i.e. resource exists).
+	tr.SetConditions(xpv1.Available())
+
+	plan, err := e.tf.Plan(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, "cannot check resource status")
+	}
+
+	// TODO(muvaf): Handle connection details.
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        plan.UpToDate,
+		ResourceLateInitialized: lateInited,
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg xpresource.Managed) (managed.ExternalCreation, error) {
+	res, err := e.Update(ctx, mg)
+	return managed.ExternalCreation{ConnectionDetails: res.ConnectionDetails}, err
+}
+
+func (e *external) Update(ctx context.Context, mg xpresource.Managed) (managed.ExternalUpdate, error) {
+	if e.async {
+		return managed.ExternalUpdate{}, errors.Wrap(e.tf.ApplyAsync(), "cannot start async apply")
+	}
+	tr, ok := mg.(resource.Terraformed)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errUnexpectedObject)
+	}
+	res, err := e.tf.Apply(ctx)
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, "cannot create")
+	}
+	attr := map[string]interface{}{}
+	if err := json.JSParser.Unmarshal(res.State.GetAttributes(), &attr); err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, "cannot unmarshal state attributes")
+	}
```

Comment:
a todo for handling connection details here as well? Not to lose one-time available connection details.